### PR TITLE
Add serialization

### PR DESCRIPTION
**Pull request by Ygg01**
Adds serialization for `bevy_input_action` by implementing a custom serializer for `inputMap.actions`. As far as I understand that is the only thing that needs serialization when preserving configuration.

Contains some minor nits.
- Upgrades examples to `bevy 0.6.0`
- Adds serialization test
- Adds serialization examples

### CODE REVIEW

File: src/serialize.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    where
+        D: Deserializer<'de>,
+    {
+        let visitor = InputMapVisitor::<T>::new();
+
+        deserializer.deserialize_map(visitor)
+    }
+}
+
+#[cfg(feature = "serialize")]
```
Comment (ndarilek) [inline src/serialize.rs:40]: Same.

File: src/serialize.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+impl<T> InputMapVisitor<T> {
+    pub(crate) fn new() -> Self {
+        InputMapVisitor {
+            marker: PhantomData,
+        }
+    }
+}
+
+#[cfg(feature = "serialize")]
```
Comment (ndarilek) [inline src/serialize.rs:53]: Same.

File: src/lib.rs [hunk 3]
```diff
@@ -13,7 +13,14 @@ use bevy::{
     prelude::*,
 };
 
+#[cfg(feature = "serialize")]
+pub use serde::{Deserialize, Deserializer, Serialize, Serializer};
```
Comment (ndarilek) [inline src/lib.rs:17]: Does this need to be `pub`?

File: src/lib.rs [hunk 4]
```diff
@@ -13,7 +13,14 @@ use bevy::{
...
+
+#[cfg(feature = "serialize")]
+mod serialize;
+
 #[derive(Clone, Debug, Default, PartialEq)]
+#[cfg_attr(feature = "serialize", derive(Serialize, Deserialize))]
 /// A single binding consisting of sets of applicable key presses or gamepad activity meant to
 /// be used in tandem.
 pub struct Binding {
```
Comment (ndarilek) [inline src/lib.rs:26]: Does this need to be `pub` or would `pub(crate)` work??

File: src/lib.rs [hunk 5]
```diff
@@ -144,9 +152,10 @@ impl Binding {
 }
 
 #[derive(Clone, Debug, Default, PartialEq)]
-/// An Action consists of many bindings of which any caount as triggering it
+#[cfg_attr(feature = "serialize", derive(Serialize, Deserialize))]
+/// An Action consists of many bindings of which any count as triggering it
 pub struct Action {
```
Comment (ndarilek) [inline src/lib.rs:157]: Likewise, `pub(crate)` here?

File: src/lib.rs [hunk 6]
```diff
@@ -144,9 +152,10 @@ impl Binding {
...
-    bindings: Vec<Binding>,
+    pub bindings: Vec<Binding>,
```
Comment (ndarilek) [inline src/lib.rs:158]: And `pub(crate)` here as well?

File: src/lib.rs [hunk 7]
```diff
@@ -224,8 +233,9 @@ impl Action {
 
 /// A Bevy resource tracking bound `Action`s (including [`KeyCode`]s, [`GamepadButtonType`]s, and
 /// [`GamepadAxisDirection`]s) generic over the application's action event type.
+#[derive(Debug)]
 pub struct InputMap<T> {
```
Comment (ndarilek) [inline src/lib.rs:237]: And `pub(crate)` here too?

File: src/lib.rs [hunk 8]
```diff
@@ -224,8 +233,9 @@ impl Action {
...
-    actions: HashMap<T, Action>,
+    pub actions: HashMap<T, Action>,
```
Comment (ndarilek) [inline src/lib.rs:238]: And `pub(crate)` here as well?

**Review by ndarilek — CHANGES_REQUESTED** [review]
Thanks a bunch! Here are some questions/suggested changes:

File: src/lib.rs [hunk 9]
```diff
@@ -72,6 +79,7 @@ impl From<Vec<GamepadButtonType>> for Binding {
 }
 
 #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+#[cfg_attr(feature = "serialize", derive(Serialize, Deserialize))]
 pub enum GamepadAxisDirection {
```
[the remark this answers is not in the document]
Comment (Ygg01) [reply src/lib.rs:83]: What do you mean? `GamepadAxisDirection` is already `pub` on master?

File: src/lib.rs (hunk 7 above)
Comment (Ygg01) [reply src/lib.rs:237]: Not sure what you mean here either

File: src/lib.rs (hunk 5 above)
Comment (Ygg01) [reply src/lib.rs:157]: You mean the `bindings`, right?

**Comment by Ygg01** [comment]
@ndarilek updates?

**Comment by ndarilek** [comment]
Thanks, merged!

**Pull request merged**